### PR TITLE
Create plugins API and add preprocessor plugin support

### PR DESCRIPTION
**Pull request by chrisbreiding**
Implements the base plugins API and preprocessor plugin support as per https://github.com/cypress-io/cypress/issues/684.

Related PRs:

Docs: https://github.com/cypress-io/cypress-documentation/pull/226
Examples: https://github.com/cypress-io/cypress-example-recipes/pull/19

### CODE REVIEW

**Comment by brian-mann** [comment]
@paulfalgout this is the PR for babel configuration and the new plugins API

File: packages/driver/test/cypress/fixtures/app.js [hunk 1]
```diff
@@ -1 +1 @@
-{foo: "bar"}
+{ 'bar' }
```
Comment (bahmutov) [inline packages/driver/test/cypress/fixtures/app.js]: why such weird javacript string here?

File: packages/server/lib/config.coffee [hunk 2]
```diff
@@ -16,8 +16,32 @@ pathHelpers = require("./util/path_helpers")
...
+  testFiles                       execTimeout
+  trashAssetsBeforeHeadlessRuns   pageLoadTimeout
+  viewportWidth                   requestTimeout
+  viewportHeight                  responseTimeout
+  videoRecording
+  videoCompression
+  videoUploadOnPasses
+  watchForFileChanges
+  waitForAnimations
+""".trim().split(/\s+/)
```
Comment (bahmutov) [inline packages/server/lib/config.coffee]: pretty good idea to have text block here. Maybe separate `.trim().split(/\s+/)` to named function that describes its purpose like `toWords`

File: packages/server/lib/config.coffee [hunk 3]
```diff
@@ -80,10 +106,12 @@ validationRules = {
   integrationFolder: v.isString
   numTestsKeptInMemory: v.isNumber
```
Comment (bahmutov) [inline packages/server/lib/config.coffee]: we could make these predicates stronger by having method `v.isPositive` to make sure user provided value is > 0

File: packages/socket/test/socket_spec.js [hunk 4]
```diff
@@ -32,10 +34,10 @@ describe('Socket', function () {
   })
 
   context('.getClientSource', function () {
-    it('returns client source as a string', function(done) {
-      const p = process.cwd() + '/node_modules/socket.io-client/dist/socket.io.js'
+    it('returns client source as a string', function (done) {
+      const p = `${process.cwd()}/node_modules/socket.io-client/dist/socket.io.js`
```
Comment (bahmutov) [inline packages/socket/test/socket_spec.js]: I think we should be more consistent about using `path.join` here

File: packages/driver/test/cypress/fixtures/app.js (hunk 1 above)
Comment (chrisbreiding) [reply packages/driver/test/cypress/fixtures/app.js]: I'm not sure what that's about. I think eslint keeps auto-formatting it for some reason.

File: packages/server/lib/config.coffee (hunk 3 above)
Comment (chrisbreiding) [reply packages/server/lib/config.coffee]: Sounds good to me. Seems out of scope for this PR though...

File: packages/socket/test/socket_spec.js (hunk 4 above)
Comment (chrisbreiding) [reply packages/socket/test/socket_spec.js]: Agreed. I think the only reason this changed is that eslint auto-fixed it.

**Comment by brian-mann** [comment]
Fixes #684

**Pull request merged**